tests/test_schema.py

```
@@ -3,9 +3,10 @@
 from nomad.client import normalize_all, parse
 
 
-def test_schema():
+def test_schema(capture_from_logger, clean_up):
```

This is what is confusing me a bit on using `capture_from_logger` as a fixture.

Do you want to capture if there is an ERROR in the logger? If so, I would suggest that: the functionality which is being tested should return something that we expect when there is an error in the logger.

Let me put an example. Imagine we define:
```python
def sum_two_positive_int(int1: int, int2: int, logger: BoundLogger) -> Optional[int]:
    if int1 < 0 or int2 < 0:
        logger.error('One of the integers is negative. Returning None.')
        return None
    return int1 + int2
```

Then testing should check:
```python
@pytest.mark.parametrize(
    'int1, int2, result',
    [
        (1, 1, 2),
        (-1, 1, None),
        (1, -1, None),
    ],
)
def test(self, int1, int2, result):
    """
    Test the `chemical_symbol` and `atomic_number` resolution for the `AtomsState` section.
    """
    assert sum_two_positive_int(int1, int2) == result
```

Thing is, if the logger stores an Error but we don't return after it, it will either crash later on the function because of some missing information or continue doing stuff that we don't want to (in the example before, the return will be 0, but maybe we don't want to run this functionality if some conditions are not satisfied).

---

@JosePizarro3  Ah, I see your point, but I think the normalization will be terminated right after it encounters an error, and that error will be logged in the logger. For example, to make a quick test, I access a missing function in `archive.m_context` in the `write_results` method of `ELNJupyterNotebook` class, and immediately after put a print statement.

```py
def write_results(self, archive: 'EntryArchive', logger: 'BoundLogger') -> None:
        ...
        if archive.m_context.NO_SUCH_FUNCTION('tmp_analysis_results.json'):
            print("Code was NOT terminated.")
        ...
```

The test fails as the logger contains the error AND the print statement is not executed: meaning that `normalize_all` was terminated right after it encountered the error.

---

In that case, I will argue that there's missing a error handling before (tho in this specific case, it was a problem of the nomad source code, and hence you had to patch it there):
```python
if not hasattr(archive.m_context, 'NO_SUCH_FUNCTION'):
    logger.error('Could not find the function `NO_SUCH_FUNCTION()` defined in archive.m_context.')
    return None
```

---

It's convenient to use `raise` statements for error handling in the plugin rather than using the `logger` object. Because this allows the program to terminate right away, and also pushes the error into the `logger` automatically. 

If we are using logger object to handle the errors in the plugin, as you said, the program will continue executing after handling the error (we want to avoid that). Plus, it is possible that the logger accumulates more than one error due to continued execution, making it difficult to debug.

---

Hence the return... but yes, both raise / logger.error + return are equivalent.